helix-term/src/ui/editor.rs

```diff
@@ -1209,6 +1210,12 @@ impl Component for EditorView {
                 let (view, doc) = current!(cx.editor);
                 view.ensure_cursor_in_view(doc, config.scrolloff);
 
+                // The focused view has changed. Consume the event and do not execute mode
+                // transition hooks.
+                if view.id != focus {
+                    return EventResult::Consumed(None);
+                }
```

Comment:
Actually, with the changes in master, these hooks should still run (if the doc exited from insert mode, we *do* want a history state).

We just need to fetch the doc matching the original `focus` rather than `current!`